const BaseCommand = require("../../../structures/BaseSlashCommand");

module.exports = class KickCmd extends BaseCommand {
  constructor(client) {
    super({
      name: "kick",
      description: "[ üõ°Ô∏è Modera√ß√£o ] Expulse algu√©m usando esse comando."
    }, {
      clientPerms: ["kickMembers"],
      userPerms: ["kickMembers"]
    }, [{
      type: 6,
      name: "usu√°rio",
      description: "O usu√°rio que ser√° expulso.",
      required: true
    }, {
      type: 3,
      name: "motivo",
      description: "O motivo pelo qual ele ser√° expulso."
    }]);
    this.client = client;
  }
  
  async exec(ctx) {
    try {
      const userid = ctx.options.find(n => n.name == "usu√°rio").value;
      const reason = ctx.options.find(n => n.name == "motivo")?.value;
      
      const member = ctx.guild.members.get(userid);
      if(member) {
        if(member.user.id == ctx.author.id) {
          return await this.handleCustomError(ctx, `Voc√™ n√£o pode se auto expulsar!`, 1, { emoji: "üòí", ephemeral: true });
        }
        if(member.user.id == this.client.user.id) {
          return await this.handleCustomError(ctx, "Voc√™ n√£o pode me expulsar!", 1, { emoji: "üò†", epheneral: true });
        }
        if(member.user.id == ctx.guild.ownerID) {
          return await this.handleCustomError(ctx, "Voc√™ n√£o pode expulsar ele, ele √© o dono daqui!", 1, { emoji: "üò®", ephemeral: true });
        }
        
        let roles = [];
        let a_roles = [];
        member.roles.forEach((id) => {
          roles.push(ctx.guild.roles.get(id));
        });
        ctx.member.roles.forEach((id) => {
          a_roles.push(ctx.guild.roles.get(id));
        });
        roles.sort((a, b) => b.position - a.position);
        a_roles.sort((a, b) => b.position - a.position);
        
        if(roles[0]?.position >= a_roles[0]?.position) {
          return this.handleCustomError(ctx, "Voc√™ n√£o pode banir ele, o cargo dele est√° acima ou na mesma posi√ß√£o que o seu cargo!", 1, { emoji: "üò®", ephemeral: true })
        }
        
        try {
          await member.kick(reason ? reason : "N√£o definido.");
          await this.handleSuccess(ctx, `\`${member.user.username}\` foi expulso com sucesso!`);
        } catch(_) {
          this.client.logger.error(`Erro ao expulsar usu√°rio: ${_}`, "SlashCommandManager");
          await this.handleCustomError(ctx, `Houve um erro ao expulsar o usu√°rio: \`${_}\``, 1, { ephemeral: true })
        }
      } else {
        await this.handleCustomError(ctx, "Membro n√£o encontrado.");
      }
    } catch(_) {
      console.log(_);
    }
  }
}